ros2_control_demo_hardware/src/diffbot_system.cpp

```diff
@@ -0,0 +1,222 @@
+        "Joint '%s' have '%s' as second state interface. '%s' expected.",
+        joint.name.c_str(), joint.state_interfaces[1].name.c_str(),
+        hardware_interface::HW_IF_VELOCITY);
+      return return_type::ERROR;
+    }
+  }
+
+  status_ = hardware_interface::status::CONFIGURED;
+  return return_type::OK;
+}
+
+std::vector<hardware_interface::StateInterface>
+DiffBotSystemHardware::export_state_interfaces()
+{
+  std::vector<hardware_interface::StateInterface> state_interfaces;
+  for (uint i = 0; i < info_.joints.size(); i++) {
+    state_interfaces.emplace_back(
+      hardware_interface::StateInterface(
+        info_.joints[i].name, hardware_interface::HW_IF_POSITION, &hw_states_[i]));
+    state_interfaces.emplace_back(
+      hardware_interface::StateInterface(
+        info_.joints[i].name, hardware_interface::HW_IF_VELOCITY, &hw_states_[i]));
+  }
+
+  return state_interfaces;
+}
+
+std::vector<hardware_interface::CommandInterface>
+DiffBotSystemHardware::export_command_interfaces()
+{
+  std::vector<hardware_interface::CommandInterface> command_interfaces;
+  for (uint i = 0; i < info_.joints.size(); i++) {
+    command_interfaces.emplace_back(
+      hardware_interface::CommandInterface(
+        info_.joints[i].name, hardware_interface::HW_IF_VELOCITY, &hw_commands_[i]));
+  }
+
+  return command_interfaces;
+}
+
+
+return_type DiffBotSystemHardware::start()
+{
+  RCLCPP_INFO(
+    rclcpp::get_logger("DiffBotSystemHardware"),
+    "Starting ...please wait...");
+
+  for (int i = 0; i <= hw_start_sec_; i++) {
+    rclcpp::sleep_for(std::chrono::seconds(1));
+    RCLCPP_INFO(
+      rclcpp::get_logger("DiffBotSystemHardware"),
+      "%.1f seconds left...", hw_start_sec_ - i);
+  }
+
+  // set some default values
+  for (uint i = 0; i < hw_states_.size(); i++) {
+    if (std::isnan(hw_states_[i])) {
+      hw_states_[i] = 0;
+      hw_commands_[i] = 0;
+    }
+  }
+
+  status_ = hardware_interface::status::STARTED;
+
+  RCLCPP_INFO(
+    rclcpp::get_logger("DiffBotSystemHardware"),
+    "System Sucessfully started!");
+
+  return return_type::OK;
+}
+
+return_type DiffBotSystemHardware::stop()
+{
+  RCLCPP_INFO(
+    rclcpp::get_logger("DiffBotSystemHardware"),
+    "Stopping ...please wait...");
+
+  for (int i = 0; i <= hw_stop_sec_; i++) {
+    rclcpp::sleep_for(std::chrono::seconds(1));
+    RCLCPP_INFO(
+      rclcpp::get_logger("DiffBotSystemHardware"),
+      "%.1f seconds left...", hw_stop_sec_ - i);
+  }
+
+  status_ = hardware_interface::status::STOPPED;
+
+  RCLCPP_INFO(
+    rclcpp::get_logger("DiffBotSystemHardware"),
+    "System sucessfully stopped!");
+
+  return return_type::OK;
+}
+
+hardware_interface::return_type DiffBotSystemHardware::read()
+{
+  RCLCPP_INFO(
+    rclcpp::get_logger("DiffBotSystemHardware"),
+    "Reading...");
+
+  for (uint i = 0; i < hw_commands_.size(); i++) {
+    // Simulate DiffBot's movement
+    hw_states_[1] = hw_commands_[i] + (hw_states_[1] - hw_commands_[i]) / hw_slowdown_;
```

Comment:
Does this slowdown make sense when using diff drive robot? I understand that your commands are velocities? you should here probably integrate to get position state and mabye create some artificial dynamics when velocity is changed. Does this make sense to you?

Reply:
I think I didn't understand the meaning of slowdown parameter. I was assuming it was a 'time constant' for rrbot dynamics.
Yes, my commands are velocities. I've changed my dynamics equation to get the velocity state. 

I am aware that is necessary to integrate to get position state but as long as I understood, since the read function is reading from the hardware, the hardware would provide those infos (including the position state already integrated).
So, I integrated `hw_states_[0] += hw_states_[1] / 2;` with 'update_rate = 2' just to exhibit the data in radians. Otherwise, Should I calculate the current and last time inside diffbot_system? What do you recommend?